File: Python/day18/AOC.py
from time import perf_counter

def parse_input(input):
    grid = []
    for line in input:
        grid.append(list(line))
    return grid
def neighbors(grid, r, c):
    hood = [(-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (-1, -1), (0, -1), (1, -1)]
    neighbors = 0
    for dr, dc in hood:
        if r + dr < 0 or r + dr >= len(grid) or c + dc < 0 or c + dc >= len(grid[0]):
            continue
        if grid[r + dr][c + dc] == '#':
            neighbors += 1
    return neighbors

def turn_on(grid, p2=[]):
    on = []
    for r, row in enumerate(grid):
        for c, col in enumerate(row):
            n = neighbors(grid, r, c)
            if col == '#':
                if n == 2 or n == 3:
                    on.append((r, c))
            elif col == '.':
                if n == 3:
                    on.append((r, c))
    if p2:
        for r, c in p2:
            on.append((r, c))
    
    return on
def part1(input, step=4, always_on=[]):
    grid = parse_input(input)
    new_grid = []
    
    for r, c in always_on:
        grid[r][c] = '#'
        
    for _ in range(len(grid)):
        new_grid.append(['.' for _ in range(len(grid[0]))])
        
    for _ in range(step):
        on = turn_on(grid, always_on)
        for r, c in on:
            new_grid[r][c] = '#'
        grid = new_grid
        new_grid = []
        for _ in range(len(grid)):
            new_grid.append(['.' for _ in range(len(grid[0]))])
    return sum([row.count('#') for row in grid])

def part2(input, step=5):
    always_on = [(0, 0), 
                 (0, len(input[0]) - 1), 
                 (len(input) - 1, 0), 
                 (len(input) - 1, len(input[0]) - 1)
            ]
    return part1(input, step, always_on)
    
def main():
    input = open("input.txt", "r").read().splitlines()
    t1 = perf_counter()
    part1_sum = part1(input, 100)
    t2  = perf_counter()
    print(f"Part 1: {part1_sum}")
    print(f'Part 1 took {t2 - t1} seconds\n')
    
    t1 = perf_counter()
    part2_sum = part2(input, 100)
    t2 = perf_counter()
    print(f"Part 2: {part2_sum}")
    print(f'Part 2 took {t2 - t1} seconds')

if __name__ == "__main__":
    main()

File: Go/day18/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func parseInput(input []string) [][]string {
	grid := [][]string{}

	for _, line := range input {
		grid = append(grid, strings.Split(line, ""))
	}

	return grid
}

func neighborsOn(grid [][]string, x, y int) int {
	neighbors := 0
	for i := -1; i < 2; i++ {
		for j := -1; j < 2; j++ {
			if (i == 0 && j == 0) || x+i < 0 || y+j < 0 || x+i >= len(grid) || y+j >= len(grid[0]) {
				continue
			}
			if grid[x+i][y+j] == "#" {
				neighbors++
			}
		}
	}
	return neighbors
}

func turnOn(grid [][]string, alwaysOn bool) [][]string {
	if alwaysOn {
		grid[0][0] = "#"
		grid[0][len(grid[0])-1] = "#"
		grid[len(grid)-1][0] = "#"
		grid[len(grid)-1][len(grid[0])-1] = "#"
	}

	newGrid := make([][]string, len(grid))
	for i := range grid {
		newGrid[i] = make([]string, len(grid[i]))
		copy(newGrid[i], grid[i])
	}

	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == "#" {
				if neighborsOn(grid, i, j) != 2 && neighborsOn(grid, i, j) != 3 {
					newGrid[i][j] = "."
				}
			} else {
				if neighborsOn(grid, i, j) == 3 {
					newGrid[i][j] = "#"
				}
			}
		}
	}

	return newGrid
}

func part1(input []string, steps int, alwaysOn bool) int {
	grid := parseInput(input)

	for i := 0; i < steps; i++ {
		grid = turnOn(grid, alwaysOn)

		if alwaysOn {
			grid[0][0] = "#"
			grid[0][len(grid[0])-1] = "#"
			grid[len(grid)-1][0] = "#"
			grid[len(grid)-1][len(grid[0])-1] = "#"
		}
	}
	count := 0
	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == "#" {
				count++
			}
		}
	}
	return count
}

func part2(input []string, steps int) int {
	return part1(input, steps, true)
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines, 100, false)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines, 100)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
